### PR TITLE
Added ability to upload photos for new items

### PR DESCRIPTION
**Pull request by rsun0**
## Status: 
🚀 Ready for review

## Description
When adding a new item to GroceryBuddy, users can now use their camera to take a photo of the item. The photo of the item will be displayed alongside its name for all users.  
Also wrote tests for the new backend functionality.

For Homework 4, by David Wang (NetID davidew2, Github Davidw1339) and Ray Sun (NetID raysun2, Github rsun0).

### CODE REVIEW

File: backend/get_image.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import json
+import model
+
+get_image_blueprint = Blueprint("get_image", __name__)
+
+
+@get_image_blueprint.route('/get_image', methods=['GET'])
+def get_image():
+    '''
+    Body: {"upc"}
```
Comment (miajs2) [inline backend/get_image.py:14]: The get_image() function should be better documented with comments to explain exactly what it does.

File: backend/add_item.py [hunk 2]
```diff
@@ -5,14 +5,33 @@
 from datetime import datetime
 import mongoengine.errors
 from utils import Error
+import base64
+from os import environ
 
 add_item_blueprint = Blueprint("add_item", __name__)
 
 
+def add_image(image_b64, item):
```
Comment (miajs2) [inline backend/add_item.py:14]: Since the file is called add_item, it would be cleaner to remove the add_image function from this class and instead move to a separate file called add_image.

File: frontend/components/CameraView.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+              )}
+          </Camera>
+        </View>
+      );
+    }
+  }
+}
+
+
+const styles = StyleSheet.create({
```
Comment (miajs2) [inline frontend/components/CameraView.js:87]: In order to keep your files short and easy to read, you could create a separate file for your styles.

File: backend/tests/test_add_item.py [hunk 4]
```diff
@@ -89,6 +88,102 @@ def test_add_new_item_with_image(client):
...
+    store = nonexistent_item.stores[0]
+    store_name = str(store.name)
+    lat = float(store.location['lat'])
+    long_arg = float(store.location['long'])
+
+    price_data = store.prices[0]
+    price = float(price_data.price)
+    user = str(price_data.user)
+
+    rv = client.post('/item', data=json.dumps({
```
Comment (miajs2) [inline backend/tests/test_add_item.py:108]: It's unclear why you used rv as your variable name. (Could just be our lack of familiarity with the details of your project)

File: backend/tests/test_add_item.py [hunk 5]
```diff
@@ -89,6 +88,102 @@ def test_add_new_item_with_image(client):
...
+    assert new_item.image is not None
+    # Check substring in image_url
+    assert new_item.image_url.find('/get_image?upc=' + new_item.upc) != -1
+
+
+def test_add_new_item_image_over_url(client, db):
+    '''
+    Test that the image field has priority over image_url
+    '''
+    nonexistent_item = test_data.item13
```
Comment (miajs2) [inline backend/tests/test_add_item.py]: I believe you use nonexistent_item to refer to an item that has yet to be added. However, this item does exist, just not yet in the db and so a better name may be new_item.

File: backend/tests/test_add_item.py [hunk 6]
```diff
@@ -89,6 +88,102 @@ def test_add_new_item_with_image(client):
     assert new_price.downvotes == []
```
Comment (miajs2) [inline backend/tests/test_add_item.py:90]: You do not seem to have a test case for when an item is added without an image, and then later, the user wants to add an image to that existing item.

File: frontend/components/AddItemForm.js [hunk 7]
```diff
@@ -77,28 +96,34 @@ export default class AddItemForm extends React.Component {
...
+            {INPUT_FIELDS.map((fieldProps) => {
+              return <Input
+                  key={fieldProps.name}
+                  onChangeText={(textInput) => this.setState({ [fieldProps.name]: textInput })}
+                  value={this.state[fieldProps.name]}
+                  containerStyle={styles.input}
+                  {...fieldProps}
+                  />
+            })}
+            <Button
```
Comment (miajs2) [inline frontend/components/AddItemForm.js:115]: You could add a button style since all buttons seem to be raised and it would make it easier to modify the styles of buttons in the future if necessary.

File: frontend/components/CameraView.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  render() {
+    const { hasCameraPermission } = this.state;
+    // Check that permission was granted
+    if (hasCameraPermission === null) {
+      return <View />;
+    } else if (hasCameraPermission === false) {
+      return <Text>No access to camera</Text>;
+    } else {
+      return (
+        <View style={styles.container}>
```
Comment (miajs2) [inline frontend/components/CameraView.js:35]: This view is quite lengthy, so it would be useful to have comments throughout that help explain what it is doing.

File: backend/add_item.py (hunk 2 above)
Comment (rsun0) [reply backend/add_item.py:14]: The files are separated to make one file per API endpoint. For example, this add_item file provides the logic for the POST /item route used by the frontend. add_image is just a helper function that does part of the work for the same route. Thus, I think it would be best if it stays in the same file. Otherwise, we would make a new file for each function by the same logic, and one file per function would be too many files.

I understand that the file name add_item may have been confusing, since it matches the name of the route function instead of the API endpoint. However, the API endpoint names are shorter ("item" for this route), so changing the file names to match endpoint names would probably be worse (because "item.py" is much less clear than "add_item.py").

tl;dr We will not make a change for this comment because the add_image function logically belongs in the same file.

File: frontend/components/AddItemForm.js [hunk 9]
```diff
@@ -1,8 +1,9 @@
 import React from 'react';
-import { View, StyleSheet, ScrollView, KeyboardAvoidingView } from 'react-native';
+import { View, StyleSheet, ScrollView, KeyboardAvoidingView, Image } from 'react-native';
 import { Ionicons } from '@expo/vector-icons';
 import { Input, Text, Button } from 'react-native-elements';
 import { addGroceryItem } from '../utils/api';
+import { ImageManipulator } from 'expo';
```
Comment (miajs2) [inline frontend/components/AddItemForm.js:7]: This seems to be one of your longer files and so you should consider trying to condense it (see other suggestion about moving stylesheet for example.

File: backend/tests/test_get_image.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import pytest
+import test_data
+from utils import Error
+import base64
+import copy
+import json
+
+
+def test_get_valid_image(client, db):
```
Comment (miajs2) [inline backend/tests/test_get_image.py:9]: Good job on naming conventions overall, particularly with test case names.

**Comment by miajs2** [comment]
Checklist:
1. Do test cases cover both the main flow and alternate flows of the relevant use case?
2. Do test cases cover both the expected outcomes and potential errors?
3. Does the code completely implement all requirements as stated in the user stories/use case?
4. Does the code conform to the project's coding standard?
5. Is there any uncalled or unreachable code?
6. Is the code sufficiently documented with clear and concise comments?
7. Is there any duplicated code that could be condensed?
8. Are there any modules that are excessively complex and should either be restructured or split into multiple routines?
9. Are all classes and variables given meaningful, consistent, and clear names?

Feedback:
1. You do not seem to have a test case that tests if a user can successfully add an image to an already existing item. Also, the user story listed above does not include any alternate flows, so it is difficult to check whether the tests adequately cover all possible flows.
2. Test cases overall seem to cover both expected outcomes and potential errors, as far as we can tell.
3. We were unable to find the use case on your wiki that correlates to this user story. 
4. We were unable to find your project's coding standard on the wiki.
5. No, this looked good.
6. Commenting seems to be inconsistent. For backend files, most functions do have comments, although they could be improved to be made more understandable and descriptive. For front end files, comments were very sparse.
7. We did not find any duplicated code.
8. Overall, the modules seemed to be split up well so that file lengths were not too long. Our only suggestion for this was to separate out the StyleSheet into its own file.
9. Names of classes and variables are descriptive and easy to understand.

 Overall good job!

File: backend/tests/test_add_item.py (hunk 6 above)
Comment (rsun0) [reply backend/tests/test_add_item.py:90]: That's because that is impossible in the current implementation; once an item is added, an image cannot be uploaded later. Uploading an image for an existing item would be a separate user story. So we don't test this case in this user story.

File: backend/get_image.py (hunk 1 above)
Comment (rsun0) [reply backend/get_image.py:14]: Good point. More explanation has been added.

File: backend/tests/test_add_item.py (hunk 4 above)
Comment (rsun0) [reply backend/tests/test_add_item.py:108]: rv is short for return value. It is a common variable name used in Flask. A more descriptive name might be "response", but we use "response" in our test to refer to the JSON data of rv (which is what we actually care about). I was actually not sure of the meaning of rv myself (we just followed examples in Flask documentation), so thanks for teaching me something!

tl;dr: No change, common variable name in Flask

File: backend/tests/test_add_item.py (hunk 5 above)
Comment (rsun0) [reply backend/tests/test_add_item.py]: Good point. The original reason was because one of our test fixtures is named nonexistent_item (some tests never add the item from this fixture, so calling the fixture new_item would be misleading). But you're right that the name is misleading in some tests, so we made the following changes:

In the tests that use the nonexistent_item fixture and then add it, we made a new variable new_item = nonexistent_item. For tests that don't use the fixture, we simply renamed nonexistent_item to new_item.

File: backend/tests/test_get_image.py (hunk 10 above)
Comment (rsun0) [reply backend/tests/test_get_image.py:9]: Thanks!

File: frontend/components/AddItemForm.js (hunk 7 above)
Comment (davidenwang) [reply frontend/components/AddItemForm.js:115]: We currently just modify the container of a button so we have a containerStyle={styles.input} for now we don't need a separate style for buttons but will add one if the need arises

File: frontend/components/CameraView.js (hunk 8 above)
Comment (davidenwang) [reply frontend/components/CameraView.js:35]: Definitely, I can add some comments

File: frontend/components/CameraView.js (hunk 3 above)
Comment (davidenwang) [reply frontend/components/CameraView.js:87]: I agree it could shorten the length of the files but I think for our team it works better to have styles and content in the same file as they are very closely related and the styles are usually not too long.

File: frontend/components/AddItemForm.js (hunk 9 above)
Comment (davidenwang) [reply frontend/components/AddItemForm.js:7]: Hmmm moving the styles is a good suggestion but it would only remove 15 lines. Perhaps I can add some comments to make the code more readable

**Review by davidenwang — COMMENTED** [review]
Thank you guys for the feedback

**Review by williamwang7 — APPROVED** [review]
all things seem to be addressed.
